fcd/pass_argrec_remill.cpp

```diff
@@ -0,0 +1,370 @@
+                                     {"CL", "CH", "CX", "ECX", "RCX", nullptr},
+                                     {"DL", "DH", "DX", "EDX", "RDX", nullptr},
+                                     {"SIL", "SI", "ESI", "RSI", nullptr},
+                                     {"DIL", "DI", "EDI", "RDI", nullptr},
+                                     {"SPL", "SP", "ESP", "RSP", nullptr},
+                                     {"R8B", "R8W", "R8D", "R8", nullptr},
+                                     {"R9B", "R9W", "R9D", "R9", nullptr},
+                                     {nullptr}};
+
+  static std::unordered_map<const char *, const char **> resolver;
+  if (resolver.empty()) {
+    for (unsigned i = 0; aliases[i][0] != nullptr; ++i) {
+      for (unsigned j = 0; aliases[i][j] != nullptr; ++j) {
+        resolver[aliases[i][j]] = aliases[i];
+      }
+    }
+  }
+
+  std::unordered_set<const char *> result({reg});
+  auto iter = resolver.find(reg);
+  if (iter != resolver.end()) {
+    for (unsigned i = 0; iter->second[i] != nullptr; ++i) {
+      result.insert(iter->second[i]);
+    }
+  }
+
+  return result;
+}
+
+static std::unordered_set<llvm::User *> UsersOfVar(llvm::Function *func,
+                                                   llvm::Value *var) {
+  std::unordered_set<llvm::User *> result;
+  if (var->hasNUsesOrMore(2)) {
+    for (auto var_user : var->users()) {
+      if (auto addr = llvm::dyn_cast<llvm::LoadInst>(var_user)) {
+        for (auto addr_user : addr->users()) {
+          result.insert(addr_user);
+        }
+      }
+    }
+  }
+  return result;
+}
+
+template <typename T>
+static llvm::User *FirstUserOfVar(llvm::Function *func, llvm::Value *var,
+                                  T &instruction_list) {
+  auto users = UsersOfVar(func, var);
+  if (!users.empty()) {
+    for (auto &inst : instruction_list) {
+      if (users.count(&inst) > 0) {
+        return &inst;
+      }
+    }
+  }
+  return nullptr;
+}
+
+template <typename T>
+static std::pair<llvm::User *, const char *> FirstUserOfReg(
+    llvm::Function *func, const char *reg, T &instruction_list) {
+  std::unordered_map<llvm::User *, const char *> users;
+  for (auto alias : RegisterAliasSet(reg)) {
+    auto var = remill::FindVarInFunction(func, alias);
+    if (auto user = FirstUserOfVar(func, var, instruction_list)) {
+      users[user] = alias;
+    }
+  }
+
+  if (!users.empty()) {
+    for (auto &inst : instruction_list) {
+      auto it = users.find(&inst);
+      if (it != users.end()) {
+        return *it;
+      }
+    }
+  }
+  return {nullptr, nullptr};
+}
+
+static std::vector<llvm::BasicBlock *> TerminalBlocksOf(llvm::Function *func) {
+  std::vector<llvm::BasicBlock *> result;
+  for (auto &block : *func) {
+    if (llvm::isa<llvm::ReturnInst>(block.getTerminator())) {
+      result.push_back(&block);
+    }
+  }
+  return result;
+}
+
+static llvm::Type *RecoverRetType(llvm::Function *func, CallingConvention &cc) {
```

Comment:
Can you provide a short comment that gives insight into the approach. Right now, this seems to focus on blocks that have a `ret` in them, and that seems too narrow in practice. I think you need to look at all paths leading to a `ret`. You might be able to simplify your life using some existing CFG pass that restructures the function. This may need to be applied to a clone of the function.

What I would consider looking into would be an combination inside-out and outside-in approach:
 - look for all definitions of return registers that lead to a `ret`, add those to `stored_regs[func]`.
 - for every call to a `func` present in `stored_regs[func]`, check if the anything from `stored_regs[func]` is used on a path that is reachable from the `call`, then do `ret_reg[func][reg] += 1` to add confidence to that register.
 - If one function tail-calls another, then you may want to union together `stored_regs[caller] += stored_regs[callee]`.
 - In descending order of confidence, type check `ret_regs[func]` for each function.
 - Some functions in `stored_regs` are never directly called, fall back to an approach like what you have, but expand it so that it looks beyond the terminal blocks.

Reply:
Also added this comment to the function, for clarification.

Reply:
Adressed in issue #18